src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -2999,23 +2999,53 @@ private void VisitStatementsWithLocalFunctions(BoundBlock block)
             if (!TrackingRegions && !block.LocalFunctions.IsDefaultOrEmpty)
             {
                 // First visit everything else
+                var queue1 = ArrayBuilder<BoundLocalFunctionStatement>.GetInstance();
                 foreach (var stmt in block.Statements)
                 {
-                    if (stmt.Kind != BoundKind.LocalFunctionStatement)
+                    if (stmt is BoundLocalFunctionStatement localFunc)
+                    {
+                        queue1.Add(localFunc);
+                    }
+                    else
                     {
                         VisitStatement(stmt);
                     }
                 }
 
-                // Now visit the local function bodies
-                foreach (var stmt in block.Statements)
+                // Now visit the local function bodies.
+                // We first visit those we have seen called (hence we know their starting state),
+                // repeating this while seeing more call sites inside other local functions,
+                // and only then we visit the completely unreachable bodies.
+                // This avoids unnecessary passes and incorrect starting states.
+                var queue2 = ArrayBuilder<BoundLocalFunctionStatement>.GetInstance();
+                for (int c = 0; c != LocalFuncUsageCount;)
                 {
-                    if (stmt is BoundLocalFunctionStatement localFunc)
+                    c = LocalFuncUsageCount;
+                    foreach (var localFunc in queue1)
                     {
-                        TakeIncrementalSnapshot(localFunc);
-                        VisitLocalFunctionStatement(localFunc);
+                        if (HasLocalFuncUsage(localFunc.Symbol))
+                        {
+                            TakeIncrementalSnapshot(localFunc);
+                            VisitLocalFunctionStatement(localFunc);
+                        }
+                        else
+                        {
+                            queue2.Add(localFunc);
+                        }
                     }
+
+                    (queue1, queue2) = (queue2, queue1);
+                    queue2.Clear();
                 }
+
+                foreach (var localFunc in queue1)
```

Comment:
>foreach (var localFunc in queue1) [](http://example.com/codeflow?start=16&length=33)

I am not sure why remaining local functions are analyzed in one go. The first function might call the third, which might call the second. In this case, it looks like we will be treating the second function as though it doesn't have any call sites, where as it does have call sites.